### ISSUE TITLE
Runner received timeout error

### ISSUE THREAD

**Issue by Mel-Github** [aquasecurity/starboard#50]
encountered the above error when trying to scan for vulnerabilities for existing deployment objects. 

However if I created a new nginx deployment based on the document and invoked the same command, there was not error. 

[starboard_log.txt](https://github.com/aquasecurity/starboard/files/4736214/starboard_log.txt)

**Comment by danielpacak** [comment]
Hi @Mel-Github Thank you for trying Starboard and opening this issue. I think there're two things to address:

- Trivy does not support scanning old images such as nginx:1.7.9 . That's the one you tried, isn't it? If you run Trivy directly on it you'll get the following error:

  ```
  $ trivy nginx:1.7.9                                                                                                                                                                                               	 
  ...  
  unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws"
  ```
  TBH we cannot do much about it. There were similar issues opened to Trivy, but we won't fix it /cc @knqyf263 
- Starboard CLI should better propagate error messages from the underlying Pod which is running a scan job - I'll create a follow up task for that. This image is a good test case for us.

**Comment by braudel** [comment]
Hi Daniel, 
I think I see something similar. Running via starboard I get a timeout, but using trivy locally it works:

```
$ starboard find vulnerabilities ds/ingress-nginx-controller --namespace ingress-nginx
error: running scan job: runner received timeout

$ kubectl get ds ingress-nginx-controller -o jsonpath='{.spec.template.spec.containers[0].image}'
quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0

$ trivy quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0
2020-06-05T15:53:22.942Z	INFO	Detecting Alpine vulnerabilities...

quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0 (alpine 3.11.3)
=====================================================================================
Total: 3 (UNKNOWN: 1, LOW: 0, MEDIUM: 2, HIGH: 0, CRITICAL: 0)
...
```

**Comment by braudel** [comment]
Is there a way of specifying a longer timeout? I can only run polatis, but trivy, kube-hunter and kube-bench time out. 

```
$ starboard kube-hunter -v 5
I0605 17:27:21.129304    9889 runnable_job.go:47] Creating runnable job: starboard/9c9be922-ab2f-48ff-8726-72ede0ea2896
I0605 17:27:21.275124    9889 reflector.go:207] Starting reflector *v1.Job (30m0s) from pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156
I0605 17:27:21.275274    9889 reflector.go:243] Listing and watching *v1.Job from pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156
I0605 17:27:21.378700    9889 shared_informer.go:270] caches populated
I0605 17:28:11.374227    9889 streamwatcher.go:114] Unexpected EOF during watch stream event decoding: unexpected EOF
I0605 17:28:11.374360    9889 reflector.go:514] pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156: Watch close - *v1.Job total 0 items received
I0605 17:28:51.129358    9889 runner.go:65] Stopping runner on timeout
error: running kube-hunter job: runner received timeout
```

**Comment by Mel-Github** [comment]
> Hi @Mel-Github Thank you for trying Starboard and opening this issue. I think there're two things to address:
> 
> * Trivy does not support scanning old images such as nginx:1.7.9 . That's the one you tried, isn't it? If you run Trivy (in Dockerless mode) directly on it you'll get the following error:
>   ```
>   $ trivy nginx:1.7.9                                                                                                                                                                                               	 
>   ...  
>   unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws"
>   ```
>   
>   
>   TBH we cannot do much about it. There were similar issues opened to Trivy, but we won't fix it /cc @knqyf263
> * Starboard CLI should better propagate error messages from the underlying Pod which is running a scan job - I'll create a follow up task for that. This image is a good test case for us.

Thanks Daniel.. Is there any way to differentiate new or old image?

**Comment by danielpacak** [comment]
> Hi Daniel,
> I think I see something similar. Running via starboard I get a timeout, but using trivy locally it works:
> 
> ```
> $ starboard find vulnerabilities ds/ingress-nginx-controller --namespace ingress-nginx
> error: running scan job: runner received timeout
> 
> $ kubectl get ds ingress-nginx-controller -o jsonpath='{.spec.template.spec.containers[0].image}'
> quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0
> 
> $ trivy quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0
> 2020-06-05T15:53:22.942Z	INFO	Detecting Alpine vulnerabilities...
> 
> quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0 (alpine 3.11.3)
> =====================================================================================
> Total: 3 (UNKNOWN: 1, LOW: 0, MEDIUM: 2, HIGH: 0, CRITICAL: 0)
> ...
> ```
> 
> This is running on k8s 1.17.5.

I've created #52 to get more insights in case of scan job failure. Currently the user of Scanner CLI has no clue why it failed.

**Comment by danielpacak** [comment]
> > Hi @Mel-Github Thank you for trying Starboard and opening this issue. I think there're two things to address:
> > 
> > * Trivy does not support scanning old images such as nginx:1.7.9 . That's the one you tried, isn't it? If you run Trivy (in Dockerless mode) directly on it you'll get the following error:
> >   ```
> >   $ trivy nginx:1.7.9                                                                                                                                                                                               	 
> >   ...  
> >   unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws"
> >   ```
> >   
> >   
> >   TBH we cannot do much about it. There were similar issues opened to Trivy, but we won't fix it /cc @knqyf263
> > * Starboard CLI should better propagate error messages from the underlying Pod which is running a scan job - I'll create a follow up task for that. This image is a good test case for us.
> 
> Thanks Daniel.. Is there any way to identify new or old image?

Maybe you could use an experimental `docker manifest` or similar command to quickly check the manifest's MimeType and version of the specified image:

```
docker manifest inspect nginx:1.7.9
unsupported manifest media type and no default available: application/vnd.docker.distribution.manifest.v1+prettyjws
```

```
docker manifest inspect nginx:1.16
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
...
```

I wish someone could add it to `kubectl outdated` plugin maybe? :)

```
$ kubectl outdated

Image                              Current       Latest       Behind       Schema Outdated    
index.docker.io/library/nginx      1.7.9         1.19.0       82           true
index.docker.io/library/nginx      1.16          1.19.0       14
```

**Comment by danielpacak** [comment]
> Is there a way of specifying a longer timeout? I can only run polatis, but trivy, kube-hunter and kube-bench time out.
> 
> ```
> $ starboard kube-hunter -v 5
> I0605 17:27:21.129304    9889 runnable_job.go:47] Creating runnable job: starboard/9c9be922-ab2f-48ff-8726-72ede0ea2896
> I0605 17:27:21.275124    9889 reflector.go:207] Starting reflector *v1.Job (30m0s) from pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156
> I0605 17:27:21.275274    9889 reflector.go:243] Listing and watching *v1.Job from pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156
> I0605 17:27:21.378700    9889 shared_informer.go:270] caches populated
> I0605 17:28:11.374227    9889 streamwatcher.go:114] Unexpected EOF during watch stream event decoding: unexpected EOF
> I0605 17:28:11.374360    9889 reflector.go:514] pkg/mod/k8s.io/client-go@v0.19.0-alpha.3/tools/cache/reflector.go:156: Watch close - *v1.Job total 0 items received
> I0605 17:28:51.129358    9889 runner.go:65] Stopping runner on timeout
> error: running kube-hunter job: runner received timeout
> ```

I've added the `--scan-job-timeout` flag to configure timeout instead of using opinionated / hardcoded values. Try [Starboard CLI v0.2.1](https://github.com/aquasecurity/starboard/releases/tag/v0.2.1) and run kube-hunter as follows:

```
$ starboard kube-hunter --scan-job-timeout 5m
```

By default, if you don't specify the `--scan-job-timeout` flag, Starboard won't timeout a scan job:

```
$ starboard kube-hunter
```

**Comment by braudel** [comment]
> 
> I've added the `--scan-job-timeout` flag to configure timeout instead of using opinionated / hardcoded values. Try [Starboard CLI v0.2.1](https://github.com/aquasecurity/starboard/releases/tag/v0.2.1) and run kube-hunter as follows:
> 
> ```
> $ starboard kube-hunter --scan-job-timeout 5m
> ```
> 
> By default, if you don't specify the `--scan-job-timeout` flag, Starboard won't timeout a scan job:
> 
> ```
> $ starboard kube-hunter
> ```

Hi Daniel,

Thank you for enabling this. I have now figured out why it was timing out for me, the job's container is trying to run as root so my psp is blocking it:

```
$ kubectl describe po 0034998c-af8b-44a7-b360-91d26f07db80-drzrm -n starboard 
Name:         0034998c-af8b-44a7-b360-91d26f07db80-drzrm
Namespace:    starboard
Priority:     0
Node:         node-nf-2/172.23.0.16
Start Time:   Sat, 06 Jun 2020 09:06:12 +0100
Labels:       controller-uid=0cdccd69-f42a-470c-a0ad-22c276d6884e
              job-name=0034998c-af8b-44a7-b360-91d26f07db80
              starboard.resource.kind=DaemonSet
              starboard.resource.name=ingress-nginx-controller
Annotations:  container.apparmor.security.beta.kubernetes.io/0034998c-af8b-44a7-b360-91d26f07db80: runtime/default
              container.apparmor.security.beta.kubernetes.io/ingress-nginx-controller: runtime/default
              kubernetes.io/psp: restricted
              seccomp.security.alpha.kubernetes.io/pod: docker/default
Status:       Pending
IP:           172.24.66.131
IPs:
  IP:           172.24.66.131
Controlled By:  Job/0034998c-af8b-44a7-b360-91d26f07db80
Init Containers:
  0034998c-af8b-44a7-b360-91d26f07db80:
    Container ID:  
    Image:         docker.io/aquasec/trivy:0.8.0
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      trivy
    Args:
      --download-db-only
      --cache-dir
      /var/lib/trivy
    State:          Waiting
      Reason:       CreateContainerConfigError
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/lib/trivy from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-tm9rk (ro)
Containers:
  ingress-nginx-controller:
    Container ID:  
    Image:         docker.io/aquasec/trivy:0.8.0
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      trivy
    Args:
      --skip-update
      --cache-dir
      /var/lib/trivy
      --no-progress
      --quiet
      --format
      json
      quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.30.0
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/lib/trivy from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-tm9rk (ro)
Conditions:
  Type              Status
  Initialized       False 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  data:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  default-token-tm9rk:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-tm9rk
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason          Age                    From                            Message
  ----     ------          ----                   ----                            -------
  Normal   Scheduled       4m41s                  default-scheduler               Successfully assigned starboard/0034998c-af8b-44a7-b360-91d26f07db80-drzrm to node-nf-2
  Normal   Pulled          4m29s (x6 over 4m39s)  kubelet, node-nf-2  Successfully pulled image "docker.io/aquasec/trivy:0.8.0"
  Warning  Failed          4m29s (x6 over 4m39s)  kubelet, node-nf-2  Error: container has runAsNonRoot and image will run as root
  Normal   SandboxChanged  4m28s (x6 over 4m39s)  kubelet, node-nf-2  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulling         4m27s (x7 over 4m40s)  kubelet, node-nf-2  Pulling image "docker.io/aquasec/trivy:0.8.0"

```
Should I raise a separate ticket for this?

**Comment by danielpacak** [comment]
@braudel Let's have a dedicated issue for PSP and close this one as I hope the timeout issue is resolved now.

**Comment by danielpacak** [comment]
Closing as the timeout issue should be resolved by #53
